server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java

```diff
@@ -0,0 +1,383 @@
+import org.elasticsearch.transport.TransportResponseHandler;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static org.elasticsearch.common.util.concurrent.ConcurrentCollections.newConcurrentMap;
+
+public abstract class PeerFinder extends AbstractComponent {
+
+    public static final String REQUEST_PEERS_ACTION_NAME = "internal:discovery/requestpeers";
+
+    // the time between attempts to find all peers
+    public static final Setting<TimeValue> DISCOVERY_FIND_PEERS_INTERVAL_SETTING =
+        Setting.timeSetting("discovery.find_peers_interval",
+            TimeValue.timeValueMillis(1000), TimeValue.timeValueMillis(1), Setting.Property.NodeScope);
+
+    private final TimeValue findPeersDelay;
+
+    private final Object mutex = new Object();
+    private final TransportService transportService;
+    private final FutureExecutor futureExecutor;
+    private final TransportAddressConnector transportAddressConnector;
+    private final ConfiguredHostsResolver configuredHostsResolver;
+
+    private volatile long currentTerm;
+    private boolean active;
+    private DiscoveryNodes lastAcceptedNodes;
+    private final Map<TransportAddress, Peer> peersByAddress = newConcurrentMap();
+    private Optional<DiscoveryNode> leader = Optional.empty();
+
+    PeerFinder(Settings settings, TransportService transportService, FutureExecutor futureExecutor,
+               TransportAddressConnector transportAddressConnector, ConfiguredHostsResolver configuredHostsResolver) {
+        super(settings);
+        findPeersDelay = DISCOVERY_FIND_PEERS_INTERVAL_SETTING.get(settings);
+        this.transportService = transportService;
+        this.futureExecutor = futureExecutor;
+        this.transportAddressConnector = transportAddressConnector;
+        this.configuredHostsResolver = configuredHostsResolver;
+
+        transportService.registerRequestHandler(REQUEST_PEERS_ACTION_NAME, Names.GENERIC, false, false,
+            PeersRequest::new,
+            (request, channel, task) -> channel.sendResponse(handlePeersRequest(request)));
+    }
+
+    public Iterable<DiscoveryNode> getFoundPeers() {
+        synchronized (mutex) {
+            return getKnownPeers();
+        }
+    }
+
+    public void activate(final DiscoveryNodes lastAcceptedNodes) {
+        logger.trace("activating PeerFinder {}", lastAcceptedNodes);
+
+        synchronized (mutex) {
+            assert active == false;
+            active = true;
+            this.lastAcceptedNodes = lastAcceptedNodes;
+            leader = Optional.empty();
+            handleWakeUp();
+        }
+    }
+
+    public void deactivate(DiscoveryNode leader) {
+        synchronized (mutex) {
+            logger.trace("deactivating PeerFinder and setting leader to {}", leader);
+            if (active) {
```

Comment:
why make this method idempotent? Is this required somewhere?

Reply:
Not any more, see f0e7155dec79f83ec76c71c44010c07b9d928b83